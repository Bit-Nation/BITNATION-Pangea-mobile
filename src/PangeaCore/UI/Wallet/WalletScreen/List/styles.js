// @flow

import { StyleSheet } from 'react-native';

import GlobalStyles from 'pangea-common-reactnative/styles';

const styles = StyleSheet.create({
  ...GlobalStyles,
  walletListContainer: {
    flex: 1,
  },
});

export default styles;
