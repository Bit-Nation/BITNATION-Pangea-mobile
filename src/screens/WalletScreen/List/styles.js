// @flow

import { StyleSheet } from 'react-native';

import GlobalStyles from '../../../global/Styles';

const styles = StyleSheet.create({
  ...GlobalStyles,
  walletListContainer: {
    flex: 1,
  },
});

export default styles;
