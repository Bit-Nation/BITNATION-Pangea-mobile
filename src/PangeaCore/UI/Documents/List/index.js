// @flow

import React from "react";
import { connect } from "react-redux";
import { View, FlatList, Text } from "react-native";
import { Fab } from "native-base";

import LucyButton from "pangea-common-reactnative/UI/LucyButton";
import PopOverModal from "../../components/PopOverModal";

import type { Navigator } from "pangea-common-reactnative/ReactNativeNavigation-types";
import NavigatorComponent from "../../NavigatorComponent";
import BackgroundImage from "pangea-common-reactnative/UI/BackgroundImage";
import { androidNavigationButtons, screen } from "pangea-common-reactnative/Screens";
import styles from "./styles";
import FakeNavigationBar from "pangea-common-reactnative/UI/FakeNavigationBar";
import ScreenTitle from "pangea-common-reactnative/UI/ScreenTitle";
import i18n from "pangea-common/i18n";
import ImagePicker from 'react-native-image-crop-picker';
import { Alert } from 'react-native';
import { 
  openDocument, 
  startDocumentCreation, 
  updateModifiedDocumentField 
} from "@pangea/documents/documents-actions";
import Loading from "pangea-common-reactnative/UI/Loading";
import type { State as DocumentsState } from "@pangea/documents/documents-reducers";
import DocumentListItem from "pangea-common-reactnative/UI/DocumentListItem";
import { getDocument } from "@pangea/documents/documents-reducers";
import PhotoActionSheet from "pangea-common-reactnative/UI/PhotoActionSheet";

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc Flag that croping image picker
   */
  circleCropping: boolean
};

type Actions = {
  /**
   * @desc Function to open a document
   * @param id Index of the document to open
   */
  openDocument: (id: number) => void,
  /**
   * @desc Function to initiate document creation.
   * @param {string} data Base64 content of file to create.
   */
  startDocumentCreation: (data: string) => void,
  /**
   * @desc Function to change field of modified document.
   * @param field Name of the field to be changed.
   * @param value New value of the field.
   */
  changeDocumentField: (field: string, value: any) => void
};

type State = {
  /**
   * @desc Name of the modal to be shown
   */
   showModal: string
};

const LUCY_MODAL_KEY = "lucyModal";

class DocumentsListScreen extends NavigatorComponent<
  Props & DocumentsState & Actions,
  State
> {
  static navigatorButtons = { ...androidNavigationButtons };
  photoActionSheet: any;

  constructor(props) {
    super(props);

    this.state = {
      showModal: ""
    };
  }

  dismissModal = () => {
    this.setState({
      showModal: ""
    });
  };

  /**
   * @desc Function to run image picker.
   * @param {boolean} isCamera Flag if source is camera.
   * @return {void}
   */
  openPicker = async (isCamera: boolean) => {
    const options = {
      cropping: this.props.circleCropping,
      mediaType: 'photo',
      cropperCircleOverlay: this.props.circleCropping,
      compressImageQuality: 0.4,
      includeBase64: true,
    };

    try {
      const result = isCamera ?
        await ImagePicker.openCamera(options)
        :
        await ImagePicker.openPicker(options);

      if (result.data) {
        this.onNewDocumentContentChosen(result.data, result.mime);
      }
    } catch (error) {
      if (error.code !== 'E_PICKER_CANCELLED') {
        Alert.alert(i18n.t('error.noCamera'));
      }
    }
  };

  onSelectItem = id => {
    this.props.openDocument(id);
    const document = getDocument(this.props, id);
    if (document == null) return;

    this.props.navigator.showModal({
      ...screen("DOCUMENT_VIEW_SCREEN"),
      title: document.name,
      passProps: {
        accountId: id,
        onCancel: () => this.props.navigator.dismissModal()
      }
    });
  };

  onStartNewDocumentContent = () => {
    if (this.photoActionSheet == null) return;
    this.photoActionSheet.show();
  };

  onNewDocumentContentChosen = (data: string, mimeType: string) => {
    this.props.startDocumentCreation(data);
    this.props.changeDocumentField("mimeType", mimeType);
    this.props.navigator.showModal({
      ...screen("DOCUMENT_MODIFY_SCREEN"),
      passProps: {
        onCancel: () => this.props.navigator.dismissModal()
      }
    });
  };

  render() {
    return (
      <View style={styles.screenContainer}>
        <BackgroundImage />
        <FakeNavigationBar />
        <View style={styles.bodyContainer}>
          <FlatList
            renderItem={item => {
              const document = item.item;
              // @todo Add preview icon.
              return (
                <DocumentListItem
                  name={document.name}
                  onPress={id => this.onSelectItem(id)}
                  id={document.id}
                  description={document.description}
                />
              );
            }}
            keyExtractor={item => `${item.id}`}
            data={this.props.documents}
            style={styles.sectionList}
            ItemSeparatorComponent={() => <View style={styles.itemSeparator} />}
          />
        </View>
        {/* <Fab
          style={styles.floatingButton}
          position='bottomRight'
          onPress={this.onStartNewDocumentContent}
        >
          <Text>+</Text>
        </Fab>
        <PhotoActionSheet
          ref={actionSheet => (this.photoActionSheet = actionSheet)}
          onImageChosen={this.onNewDocumentContentChosen}
          title={i18n.t("screens.documentsList.actionSheetTitle")}
        /> */}
        {this.props.isFetching && <Loading />}
        <LucyButton
          onPress={() => this.setState({ showModal: LUCY_MODAL_KEY })}
        />
        <PopOverModal
          visible={this.state.showModal === LUCY_MODAL_KEY}
          onCancel={this.dismissModal}
          desText={i18n.t("screens.documentsList.actionSheetTitle")}
          options={[
            {
              text: i18n.t('screens.profile.edit.editPhotoActionSheet.photoLibrary'),
              onPress: () => {
                this.openPicker(false);
                //this.dismissModal();
              }
            },
            {
              text: i18n.t('screens.profile.edit.editPhotoActionSheet.takePhoto'),
              onPress: () => {
                this.openPicker(true);
                //this.dismissModal();
              }
            },
            {
              text: i18n.t('screens.profile.edit.editPhotoActionSheet.cancel'),
              onPress: () => {
                this.dismissModal();
              }
            }
          ]}
        />
      </View>
    );
  }
}

const mapStateToProps = state => ({
  ...state.documents
});

const mapDispatchToProps = dispatch => ({
  openDocument(id) {
    dispatch(openDocument(id));
  },
  startDocumentCreation(data) {
    dispatch(startDocumentCreation(data));
  },
  changeDocumentField(field, value) {
    dispatch(updateModifiedDocumentField(field, value));
  }
});

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(DocumentsListScreen);
