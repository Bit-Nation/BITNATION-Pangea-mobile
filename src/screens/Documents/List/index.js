// @flow

import React from 'react';
import { connect } from 'react-redux';
import {
  View,
  FlatList,
} from 'react-native';

import type { Navigator } from '../../../types/ReactNativeNavigation';
import NavigatorComponent from '../../../components/common/NavigatorComponent';
import BackgroundImage from '../../../components/common/BackgroundImage';
import { screen } from '../../../global/Screens';
import styles from './styles';
import FakeNavigationBar from '../../../components/common/FakeNavigationBar';
import ScreenTitle from '../../../components/common/ScreenTitle';
import ListItem from '../../../components/common/ListItem';
import type { Document } from '../../../types/Documents';
import i18n from '../../../global/i18n';
import { openDocument, startDocumentCreation } from '../../../actions/documents';
import Loading from '../../../components/common/Loading';
import type { State as DocumentsState } from '../../../reducers/documents';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc Array of accounts to be displayed.
   */
  documents: Array<Document>,
};

type Actions = {
  /**
   * @desc Function to open a document
   * @param id Index of the document to open
   */
  openDocument: (id: string) => void,
  /**
   * @desc Function to initiate document creation.
   */
  startDocumentCreation: () => void,
}

class DocumentsListScreen extends NavigatorComponent<Props & DocumentsState & Actions> {
  constructor(props) {
    super(props);

    this.props.navigator.setButtons({
      leftButtons: [],
      rightButtons: [],
    });
  }

  onSelectItem = (id) => {
    this.props.openDocument(id);
    this.props.navigator.showModal({
      ...screen('ENTER_PASSCODE_SCREEN'),
      passProps: {
        accountId: id,
        onCancel: () => this.props.navigator.dismissModal(),
      },
    });
  };

  render() {
    return (
      <View style={styles.screenContainer}>
        <BackgroundImage />
        <FakeNavigationBar />
        <View style={styles.bodyContainer}>
          <ScreenTitle title={i18n.t('screens.documentsList.title')} />
          <FlatList
            renderItem={(item) => {
              const document = item.item;
              // @todo Add preview icon.
              return (<ListItem
                text={document.name}
                onPress={id => this.onSelectItem(id)}
                id={document.id}
                subtitle={document.description}
              />);
            }}
            keyExtractor={item => item.id}
            data={this.props.documents}
            style={styles.sectionList}
            ItemSeparatorComponent={() => (<View style={styles.itemSeparator} />)}
          />
        </View>
        {this.props.isFetching && <Loading />}
      </View>
    );
  }
}

const mapStateToProps = state => ({
  ...state.documents,
});

const mapDispatchToProps = dispatch => ({
  openDocument(id) {
    dispatch(openDocument(id));
  },
  startDocumentCreation() {
    dispatch(startDocumentCreation());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(DocumentsListScreen);
