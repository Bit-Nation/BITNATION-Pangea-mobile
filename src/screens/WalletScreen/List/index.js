// @flow

import React from 'react';
import {
  FlatList,
  View,
  RefreshControl,
} from 'react-native';

import styles from './styles';
import Images from '../../../global/AssetsImages';
import WalletCard from '../../../components/WalletCard';
import { prettyWalletBalance } from '../../../utils/formatters';
import type { WalletType } from '../../../types/Wallet';

type Props = {
  wallets: Array<WalletType>,
  onReceivePress: (WalletType) => void,
  onSendPress: (WalletType) => void,
  onTransactionPress: (WalletType) => void,
  onRefresh: () => void,
  isRefreshing: boolean,
}

/**
 * @desc Component that renders wallet list.
 * It is used inside parent component, so doesn't contain navigation etc.
 * @return {React.Component} A component.
 * @constructor
 */
const WalletList = ({
  wallets,
  onReceivePress,
  onSendPress,
  onTransactionPress,
  onRefresh,
  isRefreshing,
}: Props) => ((
  <View style={styles.walletListContainer}>
    <FlatList
      data={wallets}
      keyExtractor={item => item.currency}
      renderItem={({ item }) => {
        const balance = prettyWalletBalance(item, item.currency);
        return (<WalletCard
          imagePath={item.currency === 'ETH' ? Images.ethereumLogo : Images.xpatLogo}
          onSendPress={() => onSendPress(item)}
          onReceivePress={() => onReceivePress(item)}
          onTransactionPress={() => onTransactionPress(item)}
          nameHeading={item.name}
          balance={balance}
        />);
      }}
      refreshControl={
        <RefreshControl
          refreshing={isRefreshing}
          onRefresh={() => onRefresh()}
        />}
    />
  </View>
));

export default WalletList;
