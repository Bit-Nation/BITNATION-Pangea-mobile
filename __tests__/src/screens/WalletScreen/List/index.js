import React from 'react';
import { shallow } from 'enzyme';

import List from '../../../../../src/screens/WalletScreen/List';

describe('List Wallet', () => {
  let ListScreen;
  const propsMock = {
    wallets: [{
      balance: '0',
      currency: 'ETH',
      ethAddress: '0xtestAddress',
    }],
    onReceivePress: jest.fn(),
    onSendPress: jest.fn(),
    // onTransactionPress: jest.fn(),
    isRefreshing: jest.fn(),
    onRefresh: jest.fn(),
  };
  test('List renders correctly', () => {
    ListScreen = shallow((
      <List {...propsMock} />
    ));
    expect(ListScreen).toMatchSnapshot();
  });
});
