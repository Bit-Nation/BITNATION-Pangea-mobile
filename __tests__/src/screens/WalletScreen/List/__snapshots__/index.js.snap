// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WalletList
    wallets={
      Array [
        Object {
          "balance": "0",
          "currency": "ETH",
          "ethAddress": "0xtestAddress",
          "name": "Test wallet",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <FlatList
        data={
          Array [
            Object {
              "balance": "0",
              "currency": "ETH",
              "ethAddress": "0xtestAddress",
              "name": "Test wallet",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        refreshControl={
          <RefreshControlMock
            onRefresh={[Function]}
            refreshing={undefined}
          />
        }
        renderItem={[Function]}
        scrollEventThrottle={50}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />,
      "style": Object {
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "data": Array [
          Object {
            "balance": "0",
            "currency": "ETH",
            "ethAddress": "0xtestAddress",
            "name": "Test wallet",
          },
        ],
        "disableVirtualization": false,
        "horizontal": false,
        "initialNumToRender": 10,
        "keyExtractor": [Function],
        "maxToRenderPerBatch": 10,
        "numColumns": 1,
        "onEndReachedThreshold": 2,
        "refreshControl": <RefreshControlMock
          onRefresh={[Function]}
          refreshing={undefined}
        />,
        "renderItem": [Function],
        "scrollEventThrottle": 50,
        "updateCellsBatchingPeriod": 50,
        "windowSize": 21,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <FlatList
          data={
            Array [
              Object {
                "balance": "0",
                "currency": "ETH",
                "ethAddress": "0xtestAddress",
                "name": "Test wallet",
              },
            ]
          }
          disableVirtualization={false}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onEndReachedThreshold={2}
          refreshControl={
            <RefreshControlMock
              onRefresh={[Function]}
              refreshing={undefined}
            />
          }
          renderItem={[Function]}
          scrollEventThrottle={50}
          updateCellsBatchingPeriod={50}
          windowSize={21}
        />,
        "style": Object {
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": Array [
            Object {
              "balance": "0",
              "currency": "ETH",
              "ethAddress": "0xtestAddress",
              "name": "Test wallet",
            },
          ],
          "disableVirtualization": false,
          "horizontal": false,
          "initialNumToRender": 10,
          "keyExtractor": [Function],
          "maxToRenderPerBatch": 10,
          "numColumns": 1,
          "onEndReachedThreshold": 2,
          "refreshControl": <RefreshControlMock
            onRefresh={[Function]}
            refreshing={undefined}
          />,
          "renderItem": [Function],
          "scrollEventThrottle": 50,
          "updateCellsBatchingPeriod": 50,
          "windowSize": 21,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
